private/bufpkg/bufmodule/targeting_module.go

```diff
@@ -51,94 +54,152 @@ func (m *targetingModule) TargetFileInfos(ctx context.Context) (fileInfos []bufm
+	// that are proto files. If all target paths proto files, we can return them first.
+	if m.targetPaths != nil {
+		for _, targetPath := range m.targetPaths {
+			if normalpath.Ext(targetPath) != ".proto" {
+				// not a .proto file, therefore must be a directory
 				potentialDirPaths = append(potentialDirPaths, targetPath)
 			} else {
-				// we have a file, therefore the targetPath was a file path
-				// add to the nonImportImageFiles if does not already exist
-				if _, ok := fileInfoPaths[targetPath]; !ok {
-					fileInfoPaths[targetPath] = struct{}{}
-					fileInfo, err := bufmoduleref.NewFileInfo(
-						objectInfo.Path(),
-						objectInfo.ExternalPath(),
-						false,
-						m.Module.getModuleIdentity(),
-						m.Module.getCommit(),
-					)
-					if err != nil {
+				objectInfo, err := sourceReadBucket.Stat(ctx, targetPath)
+				if err != nil {
+					if !storage.IsNotExist(err) {
 						return nil, err
 					}
-					fileInfos = append(fileInfos, fileInfo)
+					// we do not have a file, so even though this path ends
+					// in .proto,  this could be a directory - we need to check it
+					potentialDirPaths = append(potentialDirPaths, targetPath)
+				} else {
+					// Since all of these are specific files to include, and we've already checked
+					// for duplicated excludes, we know that this file is not excluded.
+					// We have a file, therefore the targetPath was a file path
+					// add to the nonImportImageFiles if does not already exist
+					if _, ok := fileInfoPaths[targetPath]; !ok {
+						fileInfoPaths[targetPath] = struct{}{}
+						fileInfo, err := bufmoduleref.NewFileInfo(
+							objectInfo.Path(),
+							objectInfo.ExternalPath(),
+							false,
+							m.Module.getModuleIdentity(),
+							m.Module.getCommit(),
+						)
+						if err != nil {
+							return nil, err
+						}
+						fileInfos = append(fileInfos, fileInfo)
+					}
 				}
 			}
 		}
+		if len(potentialDirPaths) == 0 {
+			// We had no potential directory paths as we were able to get
+			// an file for all targetPaths, so we can return the FileInfos now
+			// this means we do not have to do the expensive O(sourceReadBucketSize) operation
+			// to check to see if each file is within a potential directory path.
+			if !m.pathsAllowNotExistOnWalk {
+				for _, excludePath := range m.excludePaths {
+					var foundPath bool
+					if walkErr := sourceReadBucket.Walk(
+						ctx,
+						"",
+						func(objectInfo storage.ObjectInfo) error {
+							if normalpath.EqualsOrContainsPath(excludePath, objectInfo.Path(), normalpath.Relative) {
+								foundPath = true
+								// TODO: should we return a custom error here and catch that error to "break"
+								// the walk early if a path is found.
+							}
+							return nil
+						},
+					); walkErr != nil {
+						return nil, walkErr
+					}
+					if !foundPath {
+						return nil, fmt.Errorf("path %q has no matching file in the image", excludePath)
+					}
```

Comment:
With respect to the `TODO`, ideally yes. You can just return the error that you intend to return and the `Walk` will stop.
```suggestion
					if walkErr := sourceReadBucket.Walk(
						ctx,
						"",
						func(objectInfo storage.ObjectInfo) error {
							if normalpath.EqualsOrContainsPath(excludePath, objectInfo.Path(), normalpath.Relative) {
								return fmt.Errorf("path %q has no matching file in the image", excludePath)
							}
							return nil
						},
					); walkErr != nil {
						return nil, walkErr
					}
```

Reply:
Based on offline discussions, we are using a sentinel error for an early return on the walk.